flex_model/core/hook_function.py

```diff
@@ -231,35 +203,58 @@ def _unpack_layer_outputs(
         """
         treedef, leaves = flatten(outputs)
 
-        # TODO: Current behaviour is always taking the first leaf as the
-        #       activation tensor. But this should be exposed as configurable
-        #       to the user.
-        tensor, other_leaves = leaves[0], leaves[1:]
+        left_leaves, tensor, right_leaves = (
+            leaves[: self.unpack_idx],
+            leaves[self.unpack_idx],
+            leaves[self.unpack_idx + 1 :],
+        )
         assert tensor is not None
 
-        def _repack(_treedef, _leaves, _edited_tensor,) -> LayerOutputs:
+        def _repack(_edited_tensor) -> LayerOutputs:
             """Pack activation tensor back into layer output container."""
-            layer_outputs = unflatten(_treedef, [_edited_tensor] + _leaves)
+            layer_outputs = unflatten(
+                treedef, left_leaves + [_edited_tensor] + right_leaves,
+            )
             return layer_outputs
 
-        return tensor, partial(_repack, treedef, other_leaves)
+        return tensor, _repack
 
-    def _default_bind_tensor_to_cpu_output(self, activation: Tensor) -> None:
-        """Binds the given activation tensor to the local CPU.
+    def _concretize_offload_function(self):
+        # Safe, these don't change dring runtime.
+        using_torch_dist = torch.distributed.is_initialized
+        using_act_dist = dist.activation_parallel_is_initialized
+        in_pp_group = dist.in_pipeline_parallel_group
+        tp_world_size = dist.get_activation_tensor_parallel_world_size
 
-        Default function for binding an activation to the user-provided
-        activation dictionary.
+        def _offload_tensor_to_cpu(activation: Tensor) -> None:
+            assert self._shared_state.output_ptr is not None
 
-        :param Tensor activation: Full activation tensor to save.
+            if not using_torch_dist() or (using_act_dist() and in_pp_group()):
+                # Tensor.to() with non_blockin=True operates on shared memory
+                # buffers automatically.
+                self._shared_state.output_ptr[
+                    self.module_name
+                ] = activation.detach().to("cpu", non_blocking=True)
```

Comment:
Change output dict from `Dict[str, Tensor] -> Dict[str, List[Tensor]]` since each module can have multiple hooks, similar in format to `FlexModel.hook_functions` collection.